### PR TITLE
Add valueinfos field to FunctionProto

### PR DESCRIPTION
**Pull request by gramalingam**
### Description

* Extend FunctionProto by adding a field to store ValueInfos for variables in the function.
* Extend the helper function and parser to support this.

### Motivation and Context

This enables type/shape information to be stored for variables in a FunctionProto, just like in a GraphProto. Note that this is optional, just like in a graph. This field may store type/shape info for the inputs and outputs of a function as well. (Note that, unlike in a graph, the input/output field in a FunctionProto do not have explicit types/shapes, they are just strings, not ValueInfo.)

### CODE REVIEW

File: onnx/test/cpp/parser_test.cc [hunk 1]
```diff
@@ -363,6 +363,58 @@ f (y, z) => (w)
...
+
+TEST(ParserTest, FunctionValueInfoTest2) {
+  const char* code = R"ONNX(
+<
+  opset_import: [ "" : 10 ],
+  domain: "ai.onnx.ml",
+  doc_string: "A function test case."
+>
+f (float[N] y, float[N] z) => (float[N] w)
+<float[N] x>
```
Comment (justinchuby) [inline onnx/test/cpp/parser_test.cc:399]: Should we have examples for both single and multiple intermediate values cases?

File: docs/IR.md [hunk 2]
```diff
@@ -210,6 +211,8 @@ is explicitly included in the signature. The attribute_proto field describes att
 
 The opset_import and node fields describe the implementation of the function.
 
+The value_info field (added in IR version 10) allows a model to store type and shape information about the values used in a function, including its inputs and outputs. Note that this is optional, and ONNX allows functions to be polymorphic.
```
Comment (justinchuby) [inline docs/IR.md:214]: When the value infos do not match the runtime type/shape when a function is invoked. Is the behavior defined? Should they be ignored?

Comment (gramalingam) [reply docs/IR.md:214]: It's the same as for the main graph's value infos. I would characterize it as an error (so, no guaranteed behavior). In practice, this info can be used to perform optimizations (which may be incorrect, if incorrect value infos are supplied).

**Review by justinchuby — APPROVED** [review]

File: docs/IR.md (hunk 2 above)
Comment (xadupre) [reply docs/IR.md:214]: Should we allow multiple signatures of the same function? Should we allow to define a function supported FLOAT, DOUBLE, FLOAT?

Comment (gramalingam) [reply docs/IR.md:214]: > Should we allow multiple signatures of the same function? Should we allow to define a function supported FLOAT, DOUBLE, FLOAT?

Yes, that is useful. That is the subject of a different PR (https://github.com/onnx/onnx/pull/5011). The two are being done as separate PRs for various reasons,

**Pull request merged**